### ISSUE TITLE
make ccl object pickle-able

### ISSUE THREAD

**Issue by beckermr** [LSSTDESC/CCL#469]
A suggestion @rmjarvis.

**Comment by beckermr** [comment]
Does anyone know how to pickle the data in C in a reasonable way? AFAIK, this is really hard to do. 

Thus pickling a `pyccl.Cosmology` object would probably only be able to save the parameters and other bits of state in python. 

I am not sure if this is worth it or not. Seems like it might be for using multiprocessing.

**Comment by rmjarvis** [comment]
The normal way is to write custom `__getstate__` and `__setstate__` functions.  If present, pickle will use these for serialization.  

You can find some examples in the pre-2.0 GalSim code base.  E.g. [Angle](https://github.com/GalSim-developers/GalSim/blob/releases/1.6/galsim/angle.py#L275) is a particularly simple one. [InterpolatedImage](https://github.com/GalSim-developers/GalSim/blob/releases/1.6/galsim/interpolatedimage.py#L569) is more complicated, deleting and rebuilding some attributes.  And serializing a generic [SBProfile](https://github.com/GalSim-developers/GalSim/blob/releases/1.6/galsim/gsobject.py#L2237) object is a total hack, using a C++ function to write the `repr`, which can then be `eval`ed.

For most things, you would probably end up with something closer to the first example.

BTW, in version 2.0, we refactored all of that stuff, so that the python classes are more trivially serializable, and most of those custom functions went away.

**Comment by rmjarvis** [comment]
Oh, and the biggest place where this is useful is indeed multiprocessing.  You can't communicate between processing with anything that is not picklable.  This was the main reason we worked hard to make everything in GalSim picklable, and it was I think a good choice, despite being a pain at times.

**Comment by beckermr** [comment]
Yeah I am aware of the methods for doing this. I am worried about this data structure 

https://github.com/LSSTDESC/CCL/blob/master/include/ccl_core.h#L77

I don't want to maintain specialized code to go through each element of the struct and grab the memory associated with each gsl object. Plus IDK what is in the GSL objects. They could have pointers too, in which case the serialization would depend on GSL itself.

**Comment by beckermr** [comment]
I can rebuild all of the splines on read, but it seems better to set them to not built and let them get build automatically if they are needed.

**Comment by rmjarvis** [comment]
FWIW, the refactoring we did was basically to just have the base `__init__`s to be pure python.  And then only when we needed a c++ class to be passed to a c++ function did we initialize that attribute.  (Like `cosmo.cosmo` in CCL.)  

For non-trivial things, we persist these C++ attributes using [`lazy_property`](https://github.com/GalSim-developers/GalSim/blob/releases/2.0/galsim/utilities.py#L1382), so they only get made once, even if they are used multiple times.  But then the serialization just requires deleting any such cached objects and let them be remade on the other side of the pickle, which is a slight inefficiency, but not too bad in most cases, since large serializations also entail a runtime cost sending the bytes across the pipe.

That might be a good solution for CCL.  Like rather than create `self.parameters` in the `Parameters.__init__`, just store the arguments you would need to call the C function.  And then let `parameters` be a lazy property that makes the C object when necessary.  So:
```
    @lazy_property
    def parameters(self):
        # Create new instance of ccl_parameters object
        # Create an internal status variable; needed to check massive neutrino
        # integral.
        status = 0
        if nz_mg == -1:
            # Create ccl_parameters without modified growth
            p, status = lib.parameters_create_nu(
                Omega_c, Omega_b, Omega_k, Neff,
                w0, wa, h, norm_pk,
                n_s, bcm_log10Mc, bcm_etab, bcm_ks,
                mnu_types[mnu_type], m_nu, status)
        else:
            # Create ccl_parameters with modified growth arrays
            p, status = lib.parameters_create_nu_vec(
                Omega_c, Omega_b, Omega_k, Neff,
                w0, wa, h, norm_pk,
                n_s, bcm_log10Mc, bcm_etab, bcm_ks,
                z_mg, df_mg, mnu_types[mnu_type], m_nu, status)
        check(status)
        return p

    def __getstate__(self):
        d = self.__dict__.copy()
        d.pop('parameters',None)
        return d

    def __setstate__(self, d):
        self.__dict__ = d
```
So you still need `__getstate__` and `__setstate__`, but they are really simple, just removing any built lazy_properties from the dict.

**Comment by beckermr** [comment]
Yep this is basically what I had in mind! :)
